vassal-app/src/main/java/VASSAL/tools/FormattedString.java

```diff
@@ -133,19 +146,28 @@ public String getLocalizedText(PropertySource ps) {
   protected String getText(PropertySource ps, boolean localized) {
     final PropertySource source = (ps == null) ? defaultProperties : ps;
     try {
-      return format.evaluate(source, props, localized);
-    }
-    catch (ExpressionException e) {
-      if (source instanceof EditablePiece) {
-        ErrorDialog.dataWarning(new BadDataReport((EditablePiece) source, Resources.getString("Error.expression_error"), format.getExpression(), e));
-      }
-      else if (source instanceof AbstractConfigurable) {
-        ErrorDialog.dataWarning(new BadDataReport((AbstractConfigurable) source, Resources.getString("Error.expression_error"), format.getExpression(), e));
+      RecursionLimiter.startExecution(this);
+      try {
+        return format.evaluate(source, props, localized);
       }
-      else {
-        ErrorDialog.dataWarning(new BadDataReport(Resources.getString("Error.expression_error"), format.getExpression(), e));
+      catch (ExpressionException e) {
+        if (source instanceof EditablePiece) {
+          ErrorDialog.dataWarning(new BadDataReport((EditablePiece) source, Resources.getString("Error.expression_error"), format.getExpression(), e));
+        }
+        else if (source instanceof AbstractConfigurable) {
+          ErrorDialog.dataWarning(new BadDataReport((AbstractConfigurable) source, Resources.getString("Error.expression_error"), format.getExpression(), e));
+        }
+        else {
+          ErrorDialog.dataWarning(new BadDataReport(Resources.getString("Error.expression_error"), format.getExpression(), e));
+        }
+        return "";
       }
-      return "";
+    }
+    catch (RecursionLimitException e) {
+      return "&lt;infiniteloop&gt;";      
```

Comment:
Strictly speaking, we'd have to have a general solution to the halting problem to know that there's an infinite loop. All we actually know is that you have a very long execution consistent with being in an infinite loop.

("possible infinite loop" is both accurate and brief.)

Reply:
Oh right I think Brent wants me to return empty string but print Bad Module Data here. I was waiting for the BMD PR but now I can fix this up since it merged.